winpr/libwinpr/sspi/Kerberos/kerberos.c

```diff
@@ -253,458 +135,377 @@ static krb5_error_code krb5_prompter(krb5_context context, void* data, const cha
+	                                       pGetKeyFn, pvGetKeyArgument, phCredential, ptsExpiry);
 
 	if (principal)
-		krb5_free_principal(ctx, principal);
+		free(principal);
+	if (package)
+		free(package);
 
-	if (ctx)
-		krb5_free_context(ctx);
-
-	return ret;
+	return status;
 }
 
-static SECURITY_STATUS setupCreds(KRB_CONTEXT* context, SSPI_CREDENTIALS* credentials)
+static SECURITY_STATUS SEC_ENTRY kerberos_FreeCredentialsHandle(PCredHandle phCredential)
 {
-	SEC_WINPR_KERBEROS_SETTINGS* kerbSettings = credentials->kerbSettings;
-	sspi_gss_OID_set_desc krb5_set = { 1, SSPI_GSS_C_SPNEGO_KRB5 };
-	sspi_gss_key_value_element_desc ccache_setting = { "ccache", NULL };
-	sspi_gss_key_value_set_desc store = { 0, &ccache_setting };
-	char* file_colon_cache = NULL;
-	OM_uint32 ret, minor;
+	UINT32 minor;
+	sspi_gss_cred_id_t credentials;
 
-	if (kerbSettings && kerbSettings->cache)
-	{
-		size_t len = _snprintf(NULL, 0, "FILE:%s", kerbSettings->cache);
-		file_colon_cache = malloc(len + 1);
-		_snprintf(file_colon_cache, len + 1, "FILE:%s", kerbSettings->cache);
-		store.count = 1;
-		ccache_setting.value = file_colon_cache;
-	}
+	if (!phCredential)
+		return SEC_E_INVALID_HANDLE;
 
-	WLog_DBG(TAG, "acquiring credentials");
-	ret = sspi_gss_acquire_cred_from(&minor, SSPI_GSS_C_NO_NAME, SSPI_GSS_C_INDEFINITE, &krb5_set,
-	                                 SSPI_GSS_C_INITIATE, &store, &(context->cred), NULL, NULL);
-	if (SSPI_GSS_ERROR(ret))
-	{
+	credentials = sspi_SecureHandleGetLowerPointer(phCredential);
 
-		BOOL isNoCreds = !!(ret & SSPI_GSS_S_NO_CRED);
-		if (isNoCreds)
-		{
-			/* Then let's try to acquire credentials using login and password,
-			 * and only those two, means not with a smartcard.
-			 * If we use smartcard-logon, the credentials have already
-			 * been acquired by pkinit process. If not, returned error previously.
-			 */
-			if (init_creds(credentials->kerbSettings, credentials->identity.User,
-			               credentials->identity.UserLength, credentials->identity.Domain,
-			               credentials->identity.DomainLength, credentials->identity.Password,
-			               credentials->identity.PasswordLength))
-				return SEC_E_NO_CREDENTIALS;
-
-			WLog_INFO(TAG, "Authenticated to Kerberos v5 via login/password");
-			ret = sspi_gss_acquire_cred_from(&minor, SSPI_GSS_C_NO_NAME, SSPI_GSS_C_INDEFINITE, &krb5_set,
-			                               SSPI_GSS_C_INITIATE, &store, &(context->cred), NULL, NULL);
+	if (credentials)
+		sspi_gss_release_cred(&minor, &credentials);
 
-		}
+	sspi_SecureHandleInvalidate(phCredential);
+	return SEC_E_OK;
+}
 
+static SECURITY_STATUS SEC_ENTRY kerberos_QueryCredentialsAttributesW(PCredHandle phCredential,
+                                                                      ULONG ulAttribute,
+                                                                      void* pBuffer)
+{
+	if (ulAttribute == SECPKG_CRED_ATTR_NAMES)
```

Comment:
Can we change that to a `switch()` statement ?

Reply:
As it stands the function doesn't actually do anything in any case, but I could do that (not my code by the way)